library/entity/multipart/src/main/java/org/quiltmc/qsl/multipart/api/AbstractEntityPart.java

```diff
@@ -0,0 +1,276 @@
+/*
+ * Copyright 2022 QuiltMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.qsl.multipart.api;
+
+import net.minecraft.entity.Entity;
+import net.minecraft.entity.EntityDimensions;
+import net.minecraft.entity.EntityPose;
+import net.minecraft.entity.boss.dragon.EnderDragonPart;
+import net.minecraft.entity.damage.DamageSource;
+import net.minecraft.nbt.NbtCompound;
+import net.minecraft.network.Packet;
+import net.minecraft.util.math.Quaternion;
+import net.minecraft.util.math.Vec3d;
+import net.minecraft.util.math.Vec3f;
+
+/**
+ * A partial implementation of an {@link EntityPart} with the most common methods implemented.
+ *
+ * @param <E> the {@link Entity} that owns this {@link EntityPart}
+ * @see EnderDragonPart
+ */
+public abstract class AbstractEntityPart<E extends Entity> extends Entity implements EntityPart<E> {
+	private final E owner;
+	private EntityDimensions partDimensions;
+	private float widthRatio = 1.0f;
+	private float heightRatio = 1.0f;
+	private Vec3d relativePosition = Vec3d.ZERO;
+	private Vec3d pivot = Vec3d.ZERO;
+
+	public AbstractEntityPart(E owner, float width, float height) {
+		super(owner.getType(), owner.world);
+		this.partDimensions = EntityDimensions.changing(width, height);
+		this.calculateDimensions();
+		this.owner = owner;
+	}
+
+	@Override
+	public E getOwner() {
+		return this.owner;
+	}
+
+	@Override
+	protected void initDataTracker() {
```

Comment:
Pulling a sneaky one, I see (although yeah, it's possibly an auto-formatter, which makes sense); `{ }`

Reply:
Auto-formatter doesn't like those at all.
I tried and tried to fix that but it just doesn't want to.

Reply:
Hm, in order to compensate for that, perhaps Checkstyle should be configured so when the auto-formatter kicks in, red squiggly lines would appear; I'll investigate that later

Reply:
I found the setting, we're saved.